File: 2402-meeting-rooms-iii/2402-meeting-rooms-iii.cpp
bool myCmp(vector<int>& a,vector<int>& b) {
    return a[0] < b[0];
}

class Solution {
public:
    typedef pair<long, int> P;
    int mostBooked(int n, vector<vector<int>>& meetings) {
        sort(meetings.begin(), meetings.end(), myCmp);
        long time = 0;
        priority_queue<int, vector<int>, greater<int>> freeRooms;
        for (int i = 0; i < n; i++)
            freeRooms.push(i);
        priority_queue<P, vector<P>, greater<P>> endTimesOfOccupiedRooms;
        map<int, int> mp;
        int i = 0;
        while (i < meetings.size()) {
            // print(freeRooms,endTimesOfOccupiedRooms,mp,time,n,i,meetings);
            while (!endTimesOfOccupiedRooms.empty() &&
                   time == endTimesOfOccupiedRooms.top().first) {
                freeRooms.push(endTimesOfOccupiedRooms.top().second);
                endTimesOfOccupiedRooms.pop();
            }
            while(i<meetings.size() && meetings[i][0]<=time && !freeRooms.empty()){
                P upcomingMeet = {meetings[i][0], meetings[i][1]};
                int duration = upcomingMeet.second - upcomingMeet.first;
                if (time >= upcomingMeet.first) {
                    if (!freeRooms.empty()) {
                        int freeRoom = freeRooms.top();
                        freeRooms.pop();
                        mp[freeRoom]++;
                        endTimesOfOccupiedRooms.push({duration + time, freeRoom});
                        i++;
                    }
                }
            }
            if(!endTimesOfOccupiedRooms.empty() && freeRooms.empty())
                time=endTimesOfOccupiedRooms.top().first;
            else
                time++;
        }
        // print(freeRooms,endTimesOfOccupiedRooms,mp,time,n,i,meetings);
        int maxi = 0;
        int ans = n - 1;
        for (int i = n - 1; i >= 0; i--) {
            if (maxi <= mp[i]) {
                maxi = mp[i];
                ans = i;
            }
        }
        return ans;
    }

    void print(priority_queue<int, vector<int>, greater<int>> freeRooms,priority_queue<P, vector<P>, greater<P>> endTimesOfOccupiedRooms,map<int, int> mp,int time,int n,int i,vector<vector<int>> meetings){
        cout<<"time = "<< time<<endl;
        if(i<meetings.size())
            cout<<"Upcoming Meet : "<<meetings[i][0]<<","<<meetings[i][1]<<endl;
        else 
            cout<<"All meetings over"<<endl;
        cout<<"FreeRooms: ";
        while(!freeRooms.empty()){
            cout<<freeRooms.top()<<",";
            freeRooms.pop();
        }
        cout<<endl<<"End Time with Occupied Rooms: "<<endl;
        while(!endTimesOfOccupiedRooms.empty()){
            cout<<endTimesOfOccupiedRooms.top().first<<","<<endTimesOfOccupiedRooms.top().second<<endl;
            endTimesOfOccupiedRooms.pop();
        }
        cout<<"Map"<<endl;
        for(int i=0;i<n;i++){
            cout<<i<<"->"<<mp[i]<<endl;
        }
        cout<<endl;
    }
};

File: 2402-meeting-rooms-iii/2402-meeting-rooms-iii.java
class Solution {
    public int mostBooked(int n, int[][] meetings) {
        Arrays.sort(meetings,(int[] l, int[] r) -> {
            return l[0]-r[0];
        });

        PriorityQueue<Integer> freeRooms = new PriorityQueue<>();

        PriorityQueue<List<Integer>> pq = new PriorityQueue<>((List<Integer> lhs,List<Integer> rhs) -> {
            if(lhs.get(0)==rhs.get(0)) return lhs.get(1)-rhs.get(1);
            return lhs.get(0)-rhs.get(0);
        });

        int currTime = 0;
        Map<Integer,Integer> mp = new HashMap();
        for(int i=0;i<n;i++) freeRooms.add(i);
        
        for(int i=0;i<meetings.length;i++) {
            currTime = meetings[i][0];
            int duration = meetings[i][1] - meetings[i][0];;
            // transfer all rooms that are free
            while(!pq.isEmpty() && pq.peek().get(0)<=currTime) freeRooms.add(pq.poll().get(1));    
            if(!freeRooms.isEmpty()) {
                mp.put(freeRooms.peek(),mp.getOrDefault(freeRooms.peek(),0)+1);
                pq.add(new ArrayList<>(Arrays.asList(currTime + duration,freeRooms.poll())));
            } else if(!pq.isEmpty()) {
                int currEndTime = pq.peek().get(0);
                int room = pq.peek().get(1);
                pq.poll();
                mp.put(room,mp.getOrDefault(room,0)+1);
                pq.add(new ArrayList(Arrays.asList(currEndTime+duration,room)));
            }
        }

        int count = 0;
        int ans = n-1;
        for(int i=n-1;i>=0;i--) {
            if(count<=mp.getOrDefault(i,0)) {
                count = mp.getOrDefault(i,0);
                ans = i;
            }
        }
        return ans;
    }
}

/*
Map
room0 -> 1
room1 -> 1
room2 -> 1
room3 -> 0

currTime = 17

pq1
room3 -> 0,3
room2 -> 10,2
room1 -> 12,1
room0 -> 12,0



17  19
18  19


*/ 